server/src/main/java/org/opensearch/cluster/service/TaskBatcher.java

```diff
@@ -195,22 +195,30 @@ void runIfNotProcessed(BatchedTask updateTask) {
             }
 
             if (toExecute.isEmpty() == false) {
-                final String tasksSummary = processTasksBySource.entrySet().stream().map(entry -> {
-                    String tasks = updateTask.describeTasks(entry.getValue());
-                    return tasks.isEmpty() ? entry.getKey() : entry.getKey() + "[" + tasks + "]";
-                }).reduce((s1, s2) -> s1 + ", " + s2).orElse("");
-
+                Function<Boolean, String> taskSummaryGenerator = (longSummaryRequired) -> {
```

Comment:
I see this function is holding the reference of local map `processTasksBySource` for a longer duration (until the tasks are executed). Can we rebuild this map from `toExecute` list only when it is required for trace logging ?

The other option is to do log longSummary only here and use shortSummary everywhere else. Thoughts ?

Reply:
I think long summary should be passed to master service so that we can log with context. I think we can compute processTasksBySource in the summary generator function.

E.g.

```
                Function<Boolean, String> taskSummaryGenerator = (longSummaryRequired) -> {
                    if (longSummaryRequired == null || !longSummaryRequired) {
                        return buildShortSummary(updateTask.batchingKey);
                    }
                    final Map<String, List<BatchedTask>> processTasksBySource = new HashMap<>();
                    for(final BatchedTask task: toExecute) {
                        processTasksBySource.computeIfAbsent(task.source, s -> new ArrayList<>()).add(task);
                    }
                    return processTasksBySource.entrySet().stream().map(entry -> {
                        String tasks = updateTask.describeTasks(entry.getValue());
                        return tasks.isEmpty() ? entry.getKey() : entry.getKey() + "[" + tasks + "]";
                    }).reduce((s1, s2) -> s1 + ", " + s2).orElse("");
                };
```